cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -1730,12 +1741,16 @@ pub fn define(
     // written to the low doubleword of the register and the regiser is zero-extended to 128 bits."
     for ty in ValueType::all_lane_types().filter(|t| t.lane_bits() >= 8) {
         let instruction = scalar_to_vector.bind_vector_from_lane(ty, sse_vector_size);
-        let template = rec_frurm.opcodes(vec![0x66, 0x0f, 0x6e]); // MOVD/MOVQ
-        if ty.lane_bits() < 64 {
-            // no 32-bit encodings for 64-bit widths
-            e.enc32(instruction.clone(), template.clone());
+        if ty.is_float() {
+            e.enc_32_64_rec(instruction, rec_null_fpr, 0);
```

Comment:
If I do:

```
v0 = f32const 0x42.42
v1 = splat.f32x4 v0
v2 = extractlane 0 v1
v3 = scalar_to_vector v2
```

Then my understanding is that, because of the previous commit and this change, the high bits of v3 will be non-zeroes, in contrary to what the comment of scalar_to_vector suggests. Am I understanding correctly? If so, I think this code should be reworked, and it'd be nice to add a test.

Reply:
Yeah, good point; I think I added this because when legalizing splat I needed to get ints/booleans into an XMM register but I realized that floats already were in the right place. Perhaps the scalar_to_vector definition should change to not specify that the higher bits are zeroed?

Reply:
It's a good question: it would introduce more nondeterminism, which wouldn't be confusing if it's properly documented.

One question is, if we decided to do this, what would be the difference between insertlane 0 and scalar_to_vector? I think they'd do the same thing, conceptually, so we could:

- either remove scalar_to_vector, and require to use insertlane 0 instead. If people want to have the higher bits zeroed, they can generate a 0 constant first.
- or keep scalar_to_vector, and it would do the zeroing on behalf of the user.

For the sake of having a minimal IR, and if having a different instruction doesn't bring any new optimization opportunity (I can't think of any brought by scalar_to_vector with zeroing semantics), I think removing scalar_to_vector would be fine. @sunfishcode, any opinions here?

Reply:
Yeah, I agree that scalar_to_vector zeroing the high lanes has overlap with insertlane.

What if make `scalar_to_vector` leave nondeterministic values in the high lanes, but document it as a low-level legalization instruction, and that frontends should generally prefer `insertlane 0`? Nondeterminism is worth avoiding when we can, but there are a variety of situations where it's useful to be able to get a scalar value into a vector register, where one knows that subsequent operations won't care about the high lanes of the vector, and the extra zeroing would be needless overhead.

Reply:
Ok, I changed the documentation of `scalar_to_vector`, to look like:
```
    Scalar To Vector -- move a value out of a scalar register and into a vector register; the 
    scalar will be moved to the lowest-order bits of the vector register. Note that this 
    instruction is intended as a low-level legalization instruction and frontends should prefer 
    insertlane; on certain architectures, scalar_to_vector may zero the highest-order bits for some
    types (e.g. integers) but not for others (e.g. floats).
```
See https://github.com/CraneStation/cranelift/pull/943/commits/b2e9e09761182781a2ec0c2b50a23aa463ef22e6#diff-7c1b843a5d2e8c61f75e0d350b3f3914R2774-R2778.